Update collection creation documentation (#118)
OpenTermsArchive/docs: content/collections/create.en.md

## Set up deployment

### Define the inventory

- On your local machine
- Clone the `<collection_id>-declarations` repository

> - Update following entries in the inventory file `deployment/inventory.yml`:

Reviewer: ```suggestion
- Update the following entries in the inventory file `deployment/inventory.yml`:
```

  - `<host>` (example: `162.19.74.224`)
  - `ansible_user: <username>` (example: `debian`)

>   - `ed25519_fingerprint: <server_ssh_fingerprint>` obtained with `ssh-keyscan -t ed25519 <host>`

Reviewer 2: An example is always welcome.

### Add fingerprint to GitHub <collection_name>-declarations settings

> - Log in on GitHub using a user account with admin privileges for the `<collection_name>-declarations` repository

Reviewer: ```suggestion
- Log in on GitHub with a user account that has admin privileges for the `<collection_name>-declarations` repository
```

- Go to `https://github.com/OpenTermsArchive/<collection_name>-declarations/settings/secrets/actions`

> - Create the `SERVER_FINGERPRINT` [secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) obtained with `ssh-keyscan -t ed25519 <host>`

Reviewer 2: ```suggestion
- Create the `SERVER_FINGERPRINT` [secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) by using the previously ed25519 fingerprint generated
```

### Set up a SSH deployment key

> Configure an SSH deployment key to enable the automated deployment process via GitHub actions.

Reviewer: ```suggestion
This key will enable automated deployment via GitHub Actions.
```

#### Create the SSH key

- Connect to the server: `ssh <username>@<host>`
- Create a new deployment SSH key: `ssh-keygen -q -N "" -f ~/.ssh/ota-deploy`
- Add the public key to `authorized_keys`: `cat ~/.ssh/ota-deploy.pub >> ~/.ssh/authorized_keys`

> - Add the private key to the SSH authentication agent: `ssh-add ~/.ssh/ota-deploy` (start the SSH agent before if necessary: `eval ${ssh-agent -s}`)

Reviewer: Why do we need to add the server's private key  to the server's agent? 🤔

Author: Great question, it seems completely useless…

> Note: user must have the right to `sudo`.

Reviewer: Is that true?

Author: I don't think so, I'd say only `cat ~/.ssh/ota-deploy.pub >> ~/.ssh/authorized_keys`

#### Add key to GitHub <collection_name>-declarations repository secrets

- Log in on GitHub using a user account with admin privileges for the `<collection_name>-declarations` repository
- Go to `https://github.com/OpenTermsArchive/<collection_name>-declarations/settings/secrets/actions`
- Create the `SERVER_SSH_KEY` secret with the previously generated deployment private key

> #### Backup key

Reviewer: ```suggestion
#### Back key up
```

> - On your local machine

Reviewer: Should not be a bullet point if not an action 🙂

> - Clone the [`engine.wiki`](https://github.com/OpenTermsArchive/engine/wiki)

Reviewer: I would assume this is known since the agent is necessarily a core team member if they have access to the shared passwords :)

Author: So, you suggest to remove it?

- Open the shared passwords database `database.kdbx` with [KeePassXC](https://keepassxc.org)
- Create the `Collection: <collection_name>` folder
- Inside this folder, add an entry with the title `Deployment SSH key`
- Attach public (`ota-deploy.pub`) and private (`ota-deploy`) key files to the entry

> - Save

Reviewer 2: ```suggestion
- Save, commit and push
```

### Allow OTA-Bot to create issues and publish dataset on GitHub

#### Create a fine-grained repo-scoped token

- Log in on GitHub as OTA-Bot user
- Go to <https://github.com/settings/personal-access-tokens/new>
- Set the “Token name”: “Issue/Release on `<collection_name>` collection”
- Leave blank “Description”

> - Set the expiry date one year later

Reviewer: ```suggestion
- Set the expiry date to the maximum possible (one year later at the time of writing)
```

> - In “Resource owner”, select “OpenTermsArchive“ organization

Reviewer: ```suggestion
- In “Resource owner”, select the “OpenTermsArchive“ organization
```

- Leave blank the following text area: `Describe why your personal access token needs access to the OpenTermsArchive organization.`
- In “Repository access”, select “Only select repositories”
- In “Select repositories”, Select “`<collection_name>`-declarations” and `<collection_name>`-versions”
- In “Permissions” → “Repository permissions”, select “Contents — Access: Read and write”
- In “Permissions” → “Repository permissions”, select “Issues — Access: Read and write”
- Validate with “Generate token and request access”

#### Backup the token

- On your local machine
- Open the shared passwords database `database.kdbx` with [KeePassXC](https://keepassxc.org)
- Inside the `Collection: <collection_name>` folder, add an entry with the title `GitHub Token`
- Copy the previously generated token in the `Password` field

> - Save

Reviewer 2: ```suggestion
- Save, commit and push
```

#### Validate the token

- Log in on GitHub using a user account with admin privileges for the OpenTermsArchive organization
- Go to <https://github.com/organizations/OpenTermsArchive/settings/personal-access-token-requests>
- Select the pending request
- Approve it

### Update inventory

#### Generate vault key

- On your local machine
- Open the shared passwords database `database.kdbx` with [KeePassXC](https://keepassxc.org)
- Inside the `Collection: <collection_name>` folder, add an entry with the title `Vault key`
- In the password field, click on the icon on the right to generate a new password
- Generate a new password (without quote or backtick)

> - Save

Reviewer 2: ```suggestion
- Save, commit and push
```

> - Generate a new password (without quote or backtick)

Reviewer: ```suggestion
- Generate a new password that does not contain any quote or backtick
```

#### Create vault key file

- On your local machine
- Go to the `<collection_id>-declarations` repository
- Go to `deployment` folder
- Create a `vault.key` file
- Copy the previously generated password inside

#### Add vault key to GitHub <collection_name>-declarations settings

- Log in on GitHub using a user account with admin privileges for the `<collection_name>-declarations` repository
- Go to `https://github.com/OpenTermsArchive/<collection_name>-declarations/settings/secrets/actions`
- Create the `ANSIBLE_VAULT_KEY` [secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) by using the previously generated vault key

#### Encrypt token and update inventory

- On your local machine
- Go to the `<collection_id>-declarations` repository
- Go to `deployment` folder
- Encrypt token: `ansible-vault encrypt_string --name 'ota_engine_github_token' '<GitHub Token>'`

> - Update token the inventory file `deployment/inventory.yml`: `ota_engine_github_token: !vault | <encrypted GitHub Token>`

Reviewer: ```suggestion
- Update token in the inventory file `deployment/inventory.yml`: `ota_engine_github_token: !vault | <encrypted GitHub Token>`
```

### Set up an OTA-Bot SSH key specific to this collection

#### Create the SSH key

> - Create a new SSH key: `ssh-keygen -t ed25519 -C bot@opentermsarchive.org -P "" -f ./<collection_name>-key`

Reviewer: > if long options exist, we should use them :)

#### Encrypt private key and update inventory

- On your local machine
- Go to the `<collection_id>-declarations` repository
- Go to `deployment` folder
- Encrypt token: `ansible-vault encrypt_string --name 'ota_engine_github_bot_private_key' '<SSH PRIVATE KEY content>'`
- Update token the inventory file `deployment/inventory.yml`: `ota_engine_github_bot_private_key: !vault | <encrypted private key>`

> #### Backup key

Reviewer: ```suggestion
#### Back key up
```

- On your local machine
- Inside the [`engine.wiki`](https://github.com/OpenTermsArchive/engine/wiki)
- Open the shared passwords database `database.kdbx` with [KeePassXC](https://keepassxc.org)
- Inside the `Collection: <collection_name>` folder, add an entry with the title `OTA-Bot GitHub SSH key`

> - Attach public (`<collection_name>-key.pub`) and private `<collection_name>-key` keys file to the entry

Reviewer 2: ```suggestion
- Attach public `<collection_name>-key.pub` and private `<collection_name>-key` keys file to the entry
```

> - Save

Reviewer 2: ```suggestion
- Save, commit and push
```

#### Attach the key to OTA-Bot GitHub user

- Log in on GitHub as OTA-Bot user
- Go to <https://github.com/settings/ssh/new>
- Set the “Title”: “`<collection_name>` collection”
- Paste the public key in “Key”
- Validate with “Add SSH Key”

> ### Set up Brevo SMTP

Reviewer: ```suggestion
### Enable sending error reports over email
```

#### Create an SMTP key

Create an SMTP key to allow sending error notifications by email.

- Log in on Brevo as Open Terms Archive admin user
- Go to <https://app.brevo.com/settings/keys/smtp>
- “Generate a new SMTP key”
- Set the SMTP key name “Name your SMTP key”: `<collection_name>` collection”
- Validate with “Generate”

#### Backup key

- On your local machine
- Inside the [`engine.wiki`](https://github.com/OpenTermsArchive/engine/wiki)
- Open the shared passwords database `database.kdbx` with [KeePassXC](https://keepassxc.org)
- Inside the `Collection: <collection_name>` folder, add an entry with the title `SMTP Key`
- Copy the previously generated key in the `Password` field

> - Save

Reviewer 2: ```suggestion
- Save, commit and push
```
